[AOTI Minifier] Exclude illegal graphs from minifier search

yushangdi opened the pull request:
Summary:
Some graphs produced by the minifier graph cutter cannot be used for AOTI/export (illegal graphs), these should be considered as graphs that don't fail in the minifier, so the minifier keeps searching.


One example is the following graph, where `true_graph_0` is an fx.GraphModule. Here, export.export() would give a `UserError` with `ErrorType = UserErrorType.INVALID_OUTPUT`.

```
      # graph():
        #     %true_graph_0 : [num_users=1] = get_attr[target=true_graph_0]
        #     return (true_graph_0,)
```

This graph could be obtained from the module below:

```python
    class M(torch.nn.Module):
        def forward(self, x, flag):
            flag = flag.item()

            def true_fn(x):
                return x.clone()

            return torch.cond(flag > 0, true_fn, true_fn, [x])
 ```



So we detect such errors, and exclude them from minifier's search (consider these graphs as didn't fail).


This is ok and won't miss any actual errors, since the AOTI minifier is only designed to catch errors in the AOTI phase anyway, it is not responsible to catching export bugs.

Test Plan:
```
buck2 run  fbcode//caffe2/test/inductor:test_minifier_utils  -- -r invalid_output
```

Differential Revision: D66143487




cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang @aakhundov

henrylhtsang commented on torch/_dynamo/repro/aoti.py:
```diff
@@ -273,6 +281,29 @@ def repro_run(options, exported_program, config_patches):
+def export_for_aoti_minifier(gm, tuple_inputs) -> Optional[torch.nn.Module]:
+    # Some graphs cannot be used for AOTI/export (illegal graphs), these should be
+    # considered as graphs that don't fail in the minifier, so the minifier keeps searching.
+    # In these case, we return None. Otherwise, we return the exported graph module.
```
probably mention that this won't affect the minifier

yushangdi replied:
added now

yushangdi committed:
[AOTI Minifier] Exclude illegal graphs from minifier search (#140999)

Summary:

Some graphs cannot be used for AOTI/export (illegal graphs), these should be considered as graphs that don't fail in the minifier, so the minifier keeps searching.


One example is the following graph, where `true_graph_0` is an fx.GraphModule.

```
      # graph():
        #     %true_graph_0 : [num_users=1] = get_attr[target=true_graph_0]
        #     return (true_graph_0,)
```

Here, export.export() would give a `UserError` with `ErrorType = UserErrorType.INVALID_OUTPUT`.

So we detect such errors, and exclude them from minifier's search (consider these graphs as didn't fail).

Test Plan:
```
buck2 run  fbcode//caffe2/test/inductor:minifier_utils_cpu  -- -r invalid_output
```

Reviewed By: henrylhtsang

Differential Revision: D66143487
